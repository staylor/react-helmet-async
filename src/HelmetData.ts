import type HelmetDispatcher from './Dispatcher';
import mapStateOnServer from './server';
import type { HelmetServerState, MappedServerState } from './types';

const instances: HelmetDispatcher[] = [];

export function clearInstances() {
  instances.length = 0;
}

export interface HelmetDataType {
  instances: HelmetDispatcher[];
  context: HelmetDataContext;
}

interface HelmetDataContext {
  helmet: HelmetServerState;
}

export const isDocument = !!(
  typeof window !== 'undefined' &&
  window.document &&
  window.document.createElement
);

export default class HelmetData implements HelmetDataType {
  instances = [];
  canUseDOM = isDocument;
  context: HelmetDataContext;

  value = {
    setHelmet: (serverState: HelmetServerState) => {
      this.context.helmet = serverState;
    },
    helmetInstances: {
      get: () => (this.canUseDOM ? instances : this.instances),
      add: (instance: HelmetDispatcher) => {
        (this.canUseDOM ? instances : this.instances).push(instance);
      },
      remove: (instance: HelmetDispatcher) => {
        const index = (this.canUseDOM ? instances : this.instances).indexOf(instance);
        (this.canUseDOM ? instances : this.instances).splice(index, 1);
      },
    },
  };

  constructor(context: any, canUseDOM?: boolean) {
    this.context = context;
    this.canUseDOM = canUseDOM || false;

    if (!canUseDOM) {
      context.helmet = mapStateOnServer({
        baseTag: [],
        bodyAttributes: {},
        encodeSpecialCharacters: true,
        htmlAttributes: {},
        linkTags: [],
        metaTags: [],
        noscriptTags: [],
        scriptTags: [],
        styleTags: [],
        title: '',
        titleAttributes: {},
      } as MappedServerState);
    }
  }
}
