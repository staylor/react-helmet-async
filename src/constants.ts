export enum TAG_PROPERTIES {
  CHARSET = 'charset',
  CSS_TEXT = 'cssText',
  HREF = 'href',
  HTTPEQUIV = 'http-equiv',
  INNER_HTML = 'innerHTML',
  ITEM_PROP = 'itemprop',
  NAME = 'name',
  PROPERTY = 'property',
  REL = 'rel',
  SRC = 'src',
}

export enum ATTRIBUTE_NAMES {
  BODY = 'bodyAttributes',
  HTML = 'htmlAttributes',
  TITLE = 'titleAttributes',
}

export enum TAG_NAMES {
  BASE = 'base',
  BODY = 'body',
  HEAD = 'head',
  HTML = 'html',
  LINK = 'link',
  META = 'meta',
  NOSCRIPT = 'noscript',
  SCRIPT = 'script',
  STYLE = 'style',
  TITLE = 'title',
  FRAGMENT = 'Symbol(react.fragment)',
}

export const SEO_PRIORITY_TAGS = {
  link: { rel: ['amphtml', 'canonical', 'alternate'] },
  script: { type: ['application/ld+json'] },
  meta: {
    charset: '',
    name: ['generator', 'robots', 'description'],
    property: [
      'og:type',
      'og:title',
      'og:url',
      'og:image',
      'og:image:alt',
      'og:description',
      'twitter:url',
      'twitter:title',
      'twitter:description',
      'twitter:image',
      'twitter:image:alt',
      'twitter:card',
      'twitter:site',
    ],
  },
};

export const VALID_TAG_NAMES = Object.values(TAG_NAMES);

export const REACT_TAG_MAP = {
  accesskey: 'accessKey',
  charset: 'charSet',
  class: 'className',
  contenteditable: 'contentEditable',
  contextmenu: 'contextMenu',
  'http-equiv': 'httpEquiv',
  itemprop: 'itemProp',
  tabindex: 'tabIndex',
};

export const HTML_TAG_MAP = Object.entries(REACT_TAG_MAP).reduce(
  (carry, [key, value]) => {
    carry[value] = key;
    return carry;
  },
  {} as { [key: string]: string }
);

export const HELMET_ATTRIBUTE = 'data-rh';
