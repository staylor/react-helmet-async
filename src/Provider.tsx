import type { PropsWithChildren } from 'react';
import React, { Component } from 'react';

import HelmetData, { isDocument } from './HelmetData';
import type { HelmetServerState } from './types';

const defaultValue = {};

export const Context = React.createContext(defaultValue);

interface ProviderProps {
  context?: {
    helmet: HelmetServerState;
  };
}

export default class HelmetProvider extends Component<PropsWithChildren<ProviderProps>> {
  static canUseDOM = isDocument;

  helmetData: HelmetData;

  constructor(props: ProviderProps) {
    super(props);

    this.helmetData = new HelmetData(this.props.context || {}, HelmetProvider.canUseDOM);
  }

  render() {
    return <Context.Provider value={this.helmetData.value}>{this.props.children}</Context.Provider>;
  }
}
