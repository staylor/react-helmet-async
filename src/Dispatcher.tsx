import { Component } from 'react';
import shallowEqual from 'shallowequal';

import handleStateChangeOnClient from './client';
import mapStateOnServer from './server';
import { reducePropsToState } from './utils';
import Provider from './Provider';
import type { HelmetServerState } from './types';

export interface DispatcherContextProp {
  setHelmet: (newState: HelmetServerState) => void;
  helmetInstances: {
    get: () => HelmetDispatcher[];
    add: (helmet: HelmetDispatcher) => void;
    remove: (helmet: HelmetDispatcher) => void;
  };
}

interface DispatcherProps {
  context: DispatcherContextProp;
}

export default class HelmetDispatcher extends Component<DispatcherProps> {
  rendered = false;

  shouldComponentUpdate(nextProps: DispatcherProps) {
    return !shallowEqual(nextProps, this.props);
  }

  emitChange() {
    const { helmetInstances, setHelmet } = this.props.context;
    let serverState = null;
    const state = reducePropsToState(
      helmetInstances.get().map(instance => {
        const props = { ...instance.props };
        // @ts-ignore
        delete props.context;
        return props;
      })
    );
    if (Provider.canUseDOM) {
      handleStateChangeOnClient(state);
    } else if (mapStateOnServer) {
      serverState = mapStateOnServer(state);
    }
    // @ts-ignore
    setHelmet(serverState);
  }

  // componentWillMount will be deprecated
  // for SSR, initialize on first render
  // constructor is also unsafe in StrictMode
  init() {
    if (this.rendered) {
      return;
    }

    this.rendered = true;

    const { helmetInstances } = this.props.context;
    helmetInstances.add(this);
    this.emitChange();
  }

  render() {
    this.init();

    return null;
  }
}
