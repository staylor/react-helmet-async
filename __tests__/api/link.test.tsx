import React from 'react';

import { Helmet } from '../../src';
import { HELMET_ATTRIBUTE } from '../../src/constants';
import { render } from '../utils';

Helmet.defaultProps.defer = false;

describe('link tags', () => {
  describe('API', () => {
    it('updates link tags', () => {
      render(
        <Helmet
          link={[
            {
              href: 'http://localhost/helmet',
              rel: 'canonical',
            },
            {
              href: 'http://localhost/style.css',
              rel: 'stylesheet',
              type: 'text/css',
            },
          ]}
        />
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];

      expect(existingTags).toBeDefined();

      const filteredTags = existingTags.filter(
        tag =>
          (tag.getAttribute('href') === 'http://localhost/style.css' &&
            tag.getAttribute('rel') === 'stylesheet' &&
            tag.getAttribute('type') === 'text/css') ||
          (tag.getAttribute('href') === 'http://localhost/helmet' &&
            tag.getAttribute('rel') === 'canonical')
      );

      expect(filteredTags.length).toBeGreaterThanOrEqual(2);
    });

    it('clears all link tags if none are specified', () => {
      render(
        <Helmet
          link={[
            {
              href: 'http://localhost/helmet',
              rel: 'canonical',
            },
          ]}
        />
      );

      render(<Helmet />);

      const tagNodes = document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`);
      const existingTags = [].slice.call(tagNodes);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it("tags without 'href' or 'rel' are not accepted, even if they are valid for other tags", () => {
      // @ts-ignore
      render(<Helmet link={[{ 'http-equiv': "won't work" }]} />);

      const tagNodes = document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`);
      const existingTags = [].slice.call(tagNodes);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it("tags 'rel' and 'href' properly use 'rel' as the primary identification for this tag, regardless of ordering", () => {
      render(
        <div>
          <Helmet
            link={[
              {
                href: 'http://localhost/helmet',
                rel: 'canonical',
              },
            ]}
          />
          <Helmet
            link={[
              {
                rel: 'canonical',
                href: 'http://localhost/helmet/new',
              },
            ]}
          />
          <Helmet
            link={[
              {
                href: 'http://localhost/helmet/newest',
                rel: 'canonical',
              },
            ]}
          />
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('rel', 'canonical');
      expect(firstTag).toHaveAttribute('href', 'http://localhost/helmet/newest');
      expect(firstTag.outerHTML).toMatchSnapshot();
    });

    it("tags with rel='stylesheet' uses the href as the primary identification of the tag, regardless of ordering", () => {
      render(
        <div>
          <Helmet
            link={[
              {
                href: 'http://localhost/style.css',
                rel: 'stylesheet',
                type: 'text/css',
                media: 'all',
              },
            ]}
          />
          <Helmet
            link={[
              {
                rel: 'stylesheet',
                href: 'http://localhost/inner.css',
                type: 'text/css',
                media: 'all',
              },
            ]}
          />
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('href', 'http://localhost/style.css');
      expect(firstTag).toHaveAttribute('rel', 'stylesheet');
      expect(firstTag).toHaveAttribute('type', 'text/css');
      expect(firstTag).toHaveAttribute('media', 'all');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('rel', 'stylesheet');
      expect(secondTag).toHaveAttribute('href', 'http://localhost/inner.css');
      expect(secondTag).toHaveAttribute('type', 'text/css');
      expect(secondTag).toHaveAttribute('media', 'all');
      expect(secondTag.outerHTML).toMatchSnapshot();
    });

    it('sets link tags based on deepest nested component', () => {
      render(
        <div>
          <Helmet
            link={[
              {
                rel: 'canonical',
                href: 'http://localhost/helmet',
              },
              {
                href: 'http://localhost/style.css',
                rel: 'stylesheet',
                type: 'text/css',
                media: 'all',
              },
            ]}
          />
          <Helmet
            link={[
              {
                rel: 'canonical',
                href: 'http://localhost/helmet/innercomponent',
              },
              {
                href: 'http://localhost/inner.css',
                rel: 'stylesheet',
                type: 'text/css',
                media: 'all',
              },
            ]}
          />
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag, thirdTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags.length).toBeGreaterThanOrEqual(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('href', 'http://localhost/style.css');
      expect(firstTag).toHaveAttribute('rel', 'stylesheet');
      expect(firstTag).toHaveAttribute('type', 'text/css');
      expect(firstTag).toHaveAttribute('media', 'all');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('href', 'http://localhost/helmet/innercomponent');
      expect(secondTag).toHaveAttribute('rel', 'canonical');
      expect(secondTag.outerHTML).toMatchSnapshot();

      expect(thirdTag).toBeInstanceOf(Element);
      expect(thirdTag.getAttribute).toBeDefined();
      expect(thirdTag).toHaveAttribute('href', 'http://localhost/inner.css');
      expect(thirdTag).toHaveAttribute('rel', 'stylesheet');
      expect(thirdTag).toHaveAttribute('type', 'text/css');
      expect(thirdTag).toHaveAttribute('media', 'all');
      expect(thirdTag.outerHTML).toMatchSnapshot();
    });

    it('allows duplicate link tags if specified in the same component', () => {
      render(
        <Helmet
          link={[
            {
              rel: 'canonical',
              href: 'http://localhost/helmet',
            },
            {
              rel: 'canonical',
              href: 'http://localhost/helmet/component',
            },
          ]}
        />
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags.length).toBeGreaterThanOrEqual(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('rel', 'canonical');
      expect(firstTag).toHaveAttribute('href', 'http://localhost/helmet');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('rel', 'canonical');
      expect(secondTag).toHaveAttribute('href', 'http://localhost/helmet/component');
      expect(secondTag.outerHTML).toMatchSnapshot();
    });

    it('overrides duplicate link tags with a single link tag in a nested component', () => {
      render(
        <div>
          <Helmet
            link={[
              {
                rel: 'canonical',
                href: 'http://localhost/helmet',
              },
              {
                rel: 'canonical',
                href: 'http://localhost/helmet/component',
              },
            ]}
          />
          <Helmet
            link={[
              {
                rel: 'canonical',
                href: 'http://localhost/helmet/innercomponent',
              },
            ]}
          />
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('rel', 'canonical');
      expect(firstTag).toHaveAttribute('href', 'http://localhost/helmet/innercomponent');
      expect(firstTag.outerHTML).toMatchSnapshot();
    });

    it('overrides single link tag with duplicate link tags in a nested component', () => {
      render(
        <div>
          <Helmet
            link={[
              {
                rel: 'canonical',
                href: 'http://localhost/helmet',
              },
            ]}
          />
          <Helmet
            link={[
              {
                rel: 'canonical',
                href: 'http://localhost/helmet/component',
              },
              {
                rel: 'canonical',
                href: 'http://localhost/helmet/innercomponent',
              },
            ]}
          />
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('rel', 'canonical');
      expect(firstTag).toHaveAttribute('href', 'http://localhost/helmet/component');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('rel', 'canonical');
      expect(secondTag).toHaveAttribute('href', 'http://localhost/helmet/innercomponent');
      expect(secondTag.outerHTML).toMatchSnapshot();
    });

    it('does not render tag when primary attribute is null', () => {
      render(
        <Helmet
          link={[
            // @ts-ignore
            { rel: 'icon', sizes: '192x192', href: null },
            {
              rel: 'canonical',
              href: 'http://localhost/helmet/component',
            },
          ]}
        />
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('rel', 'canonical');
      expect(firstTag).toHaveAttribute('href', 'http://localhost/helmet/component');
      expect(firstTag.outerHTML).toMatchSnapshot();
    });
  });

  describe('Declarative API', () => {
    it('updates link tags', () => {
      render(
        <Helmet>
          <link href="http://localhost/helmet" rel="canonical" />
          <link href="http://localhost/style.css" rel="stylesheet" type="text/css" />
        </Helmet>
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];

      expect(existingTags).toBeDefined();

      const filteredTags = existingTags.filter(
        tag =>
          (tag.getAttribute('href') === 'http://localhost/style.css' &&
            tag.getAttribute('rel') === 'stylesheet' &&
            tag.getAttribute('type') === 'text/css') ||
          (tag.getAttribute('href') === 'http://localhost/helmet' &&
            tag.getAttribute('rel') === 'canonical')
      );

      expect(filteredTags.length).toBeGreaterThanOrEqual(2);
    });

    it('clears all link tags if none are specified', () => {
      render(
        <Helmet>
          <link href="http://localhost/helmet" rel="canonical" />
        </Helmet>
      );

      render(<Helmet />);

      const tagNodes = document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`);
      const existingTags = [].slice.call(tagNodes);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it("tags without 'href' or 'rel' are not accepted, even if they are valid for other tags", () => {
      render(
        <Helmet>
          <link
            // @ts-ignore
            httpEquiv="won't work"
          />
        </Helmet>
      );

      const tagNodes = document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`);
      const existingTags = [].slice.call(tagNodes);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it("tags 'rel' and 'href' properly use 'rel' as the primary identification for this tag, regardless of ordering", () => {
      render(
        <div>
          <Helmet>
            <link href="http://localhost/helmet" rel="canonical" />
          </Helmet>
          <Helmet>
            <link rel="canonical" href="http://localhost/helmet/new" />
          </Helmet>
          <Helmet>
            <link href="http://localhost/helmet/newest" rel="canonical" />
          </Helmet>
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('rel', 'canonical');
      expect(firstTag).toHaveAttribute('href', 'http://localhost/helmet/newest');
      expect(firstTag.outerHTML).toMatchSnapshot();
    });

    it("tags with rel='stylesheet' uses the href as the primary identification of the tag, regardless of ordering", () => {
      render(
        <div>
          <Helmet>
            <link href="http://localhost/style.css" rel="stylesheet" type="text/css" media="all" />
          </Helmet>
          <Helmet>
            <link rel="stylesheet" href="http://localhost/inner.css" type="text/css" media="all" />
          </Helmet>
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('href', 'http://localhost/style.css');
      expect(firstTag).toHaveAttribute('rel', 'stylesheet');
      expect(firstTag).toHaveAttribute('type', 'text/css');
      expect(firstTag).toHaveAttribute('media', 'all');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('rel', 'stylesheet');
      expect(secondTag).toHaveAttribute('href', 'http://localhost/inner.css');
      expect(secondTag).toHaveAttribute('type', 'text/css');
      expect(secondTag).toHaveAttribute('media', 'all');
      expect(secondTag.outerHTML).toMatchSnapshot();
    });

    it('sets link tags based on deepest nested component', () => {
      render(
        <div>
          <Helmet>
            <link rel="canonical" href="http://localhost/helmet" />
            <link href="http://localhost/style.css" rel="stylesheet" type="text/css" media="all" />
          </Helmet>
          <Helmet>
            <link rel="canonical" href="http://localhost/helmet/innercomponent" />
            <link href="http://localhost/inner.css" rel="stylesheet" type="text/css" media="all" />
          </Helmet>
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag, thirdTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags.length).toBeGreaterThanOrEqual(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('href', 'http://localhost/style.css');
      expect(firstTag).toHaveAttribute('rel', 'stylesheet');
      expect(firstTag).toHaveAttribute('type', 'text/css');
      expect(firstTag).toHaveAttribute('media', 'all');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('href', 'http://localhost/helmet/innercomponent');
      expect(secondTag).toHaveAttribute('rel', 'canonical');
      expect(secondTag.outerHTML).toMatchSnapshot();

      expect(thirdTag).toBeInstanceOf(Element);
      expect(thirdTag.getAttribute).toBeDefined();
      expect(thirdTag).toHaveAttribute('href', 'http://localhost/inner.css');
      expect(thirdTag).toHaveAttribute('rel', 'stylesheet');
      expect(thirdTag).toHaveAttribute('type', 'text/css');
      expect(thirdTag).toHaveAttribute('media', 'all');
      expect(thirdTag.outerHTML).toMatchSnapshot();
    });

    it('allows duplicate link tags if specified in the same component', () => {
      render(
        <Helmet>
          <link rel="canonical" href="http://localhost/helmet" />
          <link rel="canonical" href="http://localhost/helmet/component" />
        </Helmet>
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags.length).toBeGreaterThanOrEqual(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('rel', 'canonical');
      expect(firstTag).toHaveAttribute('href', 'http://localhost/helmet');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('rel', 'canonical');
      expect(secondTag).toHaveAttribute('href', 'http://localhost/helmet/component');
      expect(secondTag.outerHTML).toMatchSnapshot();
    });

    it('overrides duplicate link tags with a single link tag in a nested component', () => {
      render(
        <div>
          <Helmet>
            <link rel="canonical" href="http://localhost/helmet" />
            <link rel="canonical" href="http://localhost/helmet/component" />
          </Helmet>
          <Helmet>
            <link rel="canonical" href="http://localhost/helmet/innercomponent" />
          </Helmet>
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('rel', 'canonical');
      expect(firstTag).toHaveAttribute('href', 'http://localhost/helmet/innercomponent');
      expect(firstTag.outerHTML).toMatchSnapshot();
    });

    it('overrides single link tag with duplicate link tags in a nested component', () => {
      render(
        <div>
          <Helmet>
            <link rel="canonical" href="http://localhost/helmet" />
          </Helmet>
          <Helmet>
            <link rel="canonical" href="http://localhost/helmet/component" />
            <link rel="canonical" href="http://localhost/helmet/innercomponent" />
          </Helmet>
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('rel', 'canonical');
      expect(firstTag).toHaveAttribute('href', 'http://localhost/helmet/component');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('rel', 'canonical');
      expect(secondTag).toHaveAttribute('href', 'http://localhost/helmet/innercomponent');
      expect(secondTag.outerHTML).toMatchSnapshot();
    });

    it('does not render tag when primary attribute is null', () => {
      render(
        <Helmet>
          <link
            rel="icon"
            sizes="192x192"
            // @ts-ignore
            href={null}
          />
          <link rel="canonical" href="http://localhost/helmet/component" />
        </Helmet>
      );

      const existingTags = [...document.head.querySelectorAll(`link[${HELMET_ATTRIBUTE}]`)];
      const [firstTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('rel', 'canonical');
      expect(firstTag).toHaveAttribute('href', 'http://localhost/helmet/component');
      expect(firstTag.outerHTML).toMatchSnapshot();
    });
  });
});
