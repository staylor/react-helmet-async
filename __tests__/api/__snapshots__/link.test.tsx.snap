// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`link tags > API > allows duplicate link tags if specified in the same component 1`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet\\" data-rh=\\"true\\">"`;

exports[`link tags > API > allows duplicate link tags if specified in the same component 2`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet/component\\" data-rh=\\"true\\">"`;

exports[`link tags > API > does not render tag when primary attribute is null 1`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet/component\\" data-rh=\\"true\\">"`;

exports[`link tags > API > overrides duplicate link tags with a single link tag in a nested component 1`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet/innercomponent\\" data-rh=\\"true\\">"`;

exports[`link tags > API > overrides single link tag with duplicate link tags in a nested component 1`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet/component\\" data-rh=\\"true\\">"`;

exports[`link tags > API > overrides single link tag with duplicate link tags in a nested component 2`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet/innercomponent\\" data-rh=\\"true\\">"`;

exports[`link tags > API > sets link tags based on deepest nested component 1`] = `"<link href=\\"http://localhost/style.css\\" rel=\\"stylesheet\\" type=\\"text/css\\" media=\\"all\\" data-rh=\\"true\\">"`;

exports[`link tags > API > sets link tags based on deepest nested component 2`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet/innercomponent\\" data-rh=\\"true\\">"`;

exports[`link tags > API > sets link tags based on deepest nested component 3`] = `"<link href=\\"http://localhost/inner.css\\" rel=\\"stylesheet\\" type=\\"text/css\\" media=\\"all\\" data-rh=\\"true\\">"`;

exports[`link tags > API > tags 'rel' and 'href' properly use 'rel' as the primary identification for this tag, regardless of ordering 1`] = `"<link href=\\"http://localhost/helmet/newest\\" rel=\\"canonical\\" data-rh=\\"true\\">"`;

exports[`link tags > API > tags with rel='stylesheet' uses the href as the primary identification of the tag, regardless of ordering 1`] = `"<link href=\\"http://localhost/style.css\\" rel=\\"stylesheet\\" type=\\"text/css\\" media=\\"all\\" data-rh=\\"true\\">"`;

exports[`link tags > API > tags with rel='stylesheet' uses the href as the primary identification of the tag, regardless of ordering 2`] = `"<link rel=\\"stylesheet\\" href=\\"http://localhost/inner.css\\" type=\\"text/css\\" media=\\"all\\" data-rh=\\"true\\">"`;

exports[`link tags > Declarative API > allows duplicate link tags if specified in the same component 1`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet\\" data-rh=\\"true\\">"`;

exports[`link tags > Declarative API > allows duplicate link tags if specified in the same component 2`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet/component\\" data-rh=\\"true\\">"`;

exports[`link tags > Declarative API > does not render tag when primary attribute is null 1`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet/component\\" data-rh=\\"true\\">"`;

exports[`link tags > Declarative API > overrides duplicate link tags with a single link tag in a nested component 1`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet/innercomponent\\" data-rh=\\"true\\">"`;

exports[`link tags > Declarative API > overrides single link tag with duplicate link tags in a nested component 1`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet/component\\" data-rh=\\"true\\">"`;

exports[`link tags > Declarative API > overrides single link tag with duplicate link tags in a nested component 2`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet/innercomponent\\" data-rh=\\"true\\">"`;

exports[`link tags > Declarative API > sets link tags based on deepest nested component 1`] = `"<link href=\\"http://localhost/style.css\\" rel=\\"stylesheet\\" type=\\"text/css\\" media=\\"all\\" data-rh=\\"true\\">"`;

exports[`link tags > Declarative API > sets link tags based on deepest nested component 2`] = `"<link rel=\\"canonical\\" href=\\"http://localhost/helmet/innercomponent\\" data-rh=\\"true\\">"`;

exports[`link tags > Declarative API > sets link tags based on deepest nested component 3`] = `"<link href=\\"http://localhost/inner.css\\" rel=\\"stylesheet\\" type=\\"text/css\\" media=\\"all\\" data-rh=\\"true\\">"`;

exports[`link tags > Declarative API > tags 'rel' and 'href' properly use 'rel' as the primary identification for this tag, regardless of ordering 1`] = `"<link href=\\"http://localhost/helmet/newest\\" rel=\\"canonical\\" data-rh=\\"true\\">"`;

exports[`link tags > Declarative API > tags with rel='stylesheet' uses the href as the primary identification of the tag, regardless of ordering 1`] = `"<link href=\\"http://localhost/style.css\\" rel=\\"stylesheet\\" type=\\"text/css\\" media=\\"all\\" data-rh=\\"true\\">"`;

exports[`link tags > Declarative API > tags with rel='stylesheet' uses the href as the primary identification of the tag, regardless of ordering 2`] = `"<link rel=\\"stylesheet\\" href=\\"http://localhost/inner.css\\" type=\\"text/css\\" media=\\"all\\" data-rh=\\"true\\">"`;
